### PR TITLE
Initial change to support Streaming of trace data to DDR Memory

### PR DESCRIPTION
**Pull request by hackwa**
This change requires debug IP Layout support in xdp (in progress)

### CODE REVIEW

**Comment by maxzhen** [comment]
retest this please

File: src/runtime_src/core/include/xclperf.h [hunk 1]
```diff
@@ -409,6 +409,7 @@ typedef struct {
   unsigned int TraceID;
   unsigned char Error;
   unsigned char Reserved;
+  bool isClockTrain = false;
```
Comment (stsoe) [inline src/runtime_src/core/include/xclperf.h]: This doesn't look like "C" code.  I don't think in-class initialization is allowed in C.  xrt.h, which includes xclperf.h can be used in C programs.

File: src/runtime_src/xdp/profile/core/rt_profile.h [hunk 2]
```diff
@@ -147,7 +147,7 @@ namespace xdp {
 
     // Log device trace
     void logDeviceTrace(std::string deviceName, std::string binaryName, xclPerfMonType type,
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/core/rt_profile.h]: I know this isn't changed in this PR, but please pass strings as const std::string& when then are constants that remain unchanged in function

File: src/runtime_src/xdp/profile/device/device_intf.cpp [hunk 3]
```diff
@@ -450,7 +440,7 @@ DeviceIntf::~DeviceIntf()
        */
       std::string warnMsg = "Multiple live processes running on device. Hardware Debug and Profiling data will be unavailable for this process.";
       std::cout << warnMsg << std::endl;
-//      xrt_core::message::send(xrt_core::message::severity_level::XRT_WARNING, "XRT", warnMsg) ;
+//      xrt_core::message::send(xrt_core::message:  :severity_level::XRT_WARNING, "XRT", warnMsg) ;
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/device_intf.cpp]: ?

File: src/runtime_src/xdp/profile/device/device_intf.cpp [hunk 4]
```diff
@@ -523,4 +513,89 @@ DeviceIntf::~DeviceIntf()
     }
   }
 
+  bool DeviceIntf::initTs2mm(uint64_t bo_size)
+  {
+    xrt::device* xrtDevice = (xrt::device*)mDeviceHandle;
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/device_intf.cpp]: No c-style cast in c++ code.   and when casting, please assign to auto, why repeat the time both left and right?

File: src/runtime_src/xdp/profile/device/device_intf.cpp [hunk 5]
```diff
@@ -523,4 +513,89 @@ DeviceIntf::~DeviceIntf()
...
+  }
+
+  uint64_t DeviceIntf::getWordCountTs2mm()
+  {
+    return traceDMA->getWordCount();
+  }
+
+void* DeviceIntf::syncTraceBO(uint64_t offset, uint64_t bytes)
+{
+    xrt::device* xrtDevice = (xrt::device*)mDeviceHandle;
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/device_intf.cpp]: No c-style casts

File: src/runtime_src/xdp/profile/device/traceS2MM.cpp [hunk 6]
```diff
@@ -60,6 +107,124 @@ void TraceS2MM::showProperties()
...
+  read(TS2MM_WRITE_OFFSET_HIGH, 4, &reg_read);
+  (*outputStream) << "INFO Trace high write offset : " << reg_read << std::endl;
+  read(TS2MM_WRITTEN_LOW, 4, &reg_read);
+  (*outputStream) << "INFO Trace written low : " << reg_read << std::endl;
+  read(TS2MM_WRITTEN_HIGH, 4, &reg_read);
+  (*outputStream) << "INFO Trace written high: " << reg_read << std::dec << std::endl;
+}
 
-}   // namespace xdp
+inline void TraceS2MM::parsePacketClockTrain(uint64_t packet, uint64_t firstTimestamp, unsigned mod, xclTraceResults &result)
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/traceS2MM.cpp]: unsigned what ? this is a very bad habit used by almost everyone :-)  won't compile on windows.

File: src/runtime_src/xdp/profile/device/traceS2MM.cpp [hunk 7]
```diff
@@ -60,6 +107,124 @@ void TraceS2MM::showProperties()
...
+        << "Flags : " << static_cast<int>(result.EventFlags) << "   "
+        << "Interval : " << result.Timestamp - previousTimestamp << "   "
+        << std::endl;
+        previousTimestamp = result.Timestamp;
+    }
+}
+
+void TraceS2MM::parseTraceBuf(void* buf, uint64_t size, xclTraceResultsVector& traceVector)
+{
+    unsigned packetSizeBytes = 8;
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/traceS2MM.cpp]: unsigned what?

File: src/runtime_src/xdp/profile/device/traceS2MM.cpp [hunk 8]
```diff
@@ -60,6 +107,124 @@ void TraceS2MM::showProperties()
...
+      }
+      traceVector.mLength = tvindex;
+    } // For i < count
+    mclockTrainingdone = true;
+}
+
+  // Convert decimal to binary string
+  // NOTE: length of string is always sizeof(uint32_t) * 8
+  std::string TraceS2MM::dec2bin(uint32_t n) {
+    char result[(sizeof(uint32_t) * 8) + 1];
```
Comment (stsoe) [inline src/runtime_src/xdp/profile/device/traceS2MM.cpp]: Would it be more efficient to allocate the string up front using std::string(count, 0) then get the char* pointer using data()?  That way RVO (return value optimization) can kick in.

File: src/runtime_src/core/include/xclperf.h (hunk 1 above)
Comment (hackwa) [reply src/runtime_src/core/include/xclperf.h]: removed initialization

File: src/runtime_src/xdp/profile/core/rt_profile.h (hunk 2 above)
Comment (hackwa) [reply src/runtime_src/xdp/profile/core/rt_profile.h]: Thanks for pointing this out! I found a cleaned this up in a lot of places in our code.

File: src/runtime_src/xdp/profile/device/device_intf.cpp (hunk 4 above)
Comment (hackwa) [reply src/runtime_src/xdp/profile/device/device_intf.cpp]: fixed

File: src/runtime_src/xdp/profile/device/traceS2MM.cpp (hunk 8 above)
Comment (hackwa) [reply src/runtime_src/xdp/profile/device/traceS2MM.cpp]: updated this function

**Comment by hackwa** [comment]
retest this please

**Pull request merged**